zio-interop-cats/shared/src/main/scala/zio/interop/package.scala

```diff
@@ -34,21 +34,35 @@ package object interop {
   @inline def toEffect[F[_], R, A](rio: RIO[R, A])(implicit R: Runtime[R], F: Async[F], trace: Trace): F[A] =
     F.defer {
       val interrupted = new AtomicBoolean(true)
-      F.async[Exit[Throwable, A]] { cb =>
-        Unsafe.unsafe { implicit unsafe =>
-          val fiber          = R.unsafe.fork {
+      F.asyncCheckAttempt[Exit[Throwable, A]] { cb =>
+        F.delay {
+          implicit val unsafe: Unsafe = Unsafe.unsafe
+
+          val out = R.unsafe.runOrFork {
             signalOnNoExternalInterrupt {
               rio
             }(ZIO.succeed(interrupted.set(false)))
           }
-          fiber.unsafe
-            .addObserver(exit => cb(Right(exit)))
-          val cancelerEffect = F.delay {
-            val _ = fiber.interrupt
+          out match {
+            case Left(fiber) =>
+              val completeCb = (exit: Exit[Throwable, A]) => cb(Right(exit))
+              fiber.unsafe.addObserver(completeCb)
+              Left(Some(F.async[Unit] { cb =>
+                F.delay {
+                  val interruptCb = (_: Exit[Throwable, A]) => cb(Right(()))
+                  fiber.unsafe.addObserver(interruptCb)
+                  fiber.unsafe.removeObserver(completeCb)
+                  fiber.tellInterrupt(Cause.interrupt(fiber.id))
+                  // Allow the interruption to be interrupted
```

Comment:
Huh thanks for pointing this out. I was debating with my own self whether to add it 😬

Only reason I added it was _just in case_ CE adds support for timing out interruption. I'm happy to remove it though to make the code more readable. Wdut?

Reply:
> Could you then cancel the cancellation of the cancellation

My thinking is that some frameworks allow you to set a deadline for cancellation. But yeah I think let's remove it since it's a bit of an overkill for something that's not supported yet